File: binding/development/subarchitectures/binder/src/java/binder/abstr/BindingWorkingMemoryReader.java
package binder.abstr;


import java.util.Vector;

import binder.autogen.core.Union;
import binder.autogen.core.UnionConfiguration;
import cast.architecture.ChangeFilterFactory;
import cast.architecture.ManagedComponent;
import cast.architecture.WorkingMemoryChangeReceiver;
import cast.cdl.WorkingMemoryChange;
import cast.cdl.WorkingMemoryOperation;

/**
 * Abstract class for retrieving elements currently present in the working memory
 * of the binder
 * 
 * @author Pierre Lison
 * @version 31/08/2009
 */
public class BindingWorkingMemoryReader extends ManagedComponent {

	
	Vector<Union> currentUnions ;
	
	@Override
	public void start() {

		// if the set of possible union configurations has been updated, update the
		// monitor accordingly
		addChangeFilter(ChangeFilterFactory.createGlobalTypeFilter(UnionConfiguration.class,
				WorkingMemoryOperation.WILDCARD), new WorkingMemoryChangeReceiver() {

			public void workingMemoryChanged(WorkingMemoryChange _wmc) {
				try {
					UnionConfiguration config = 
						getMemoryEntry(_wmc.address, UnionConfiguration.class);
					extractUnionsFromConfig(config);
				}
				catch (Exception e) {
					e.printStackTrace();
				}
			} 
		});
		
		currentUnions = new Vector<Union>();
		
	}

	
	private void extractUnionsFromConfig (UnionConfiguration config) {
		currentUnions = new Vector<Union>();
		for (int i = 0; i < config.includedUnions.length ; i++) {
			currentUnions.add(config.includedUnions[i]);
		}
	}
	
	public Vector<Union> getUnions () {
		return currentUnions;
	}
	
}

File: subarchitectures/binder/branches/praised-dora-version1/src/c++/BindingWorkingMemoryReader.cpp
#include <cast/architecture/ChangeFilterFactory.hpp>
#include "BindingWorkingMemoryReader.hpp"

namespace binder {

  using namespace std;
  using namespace autogen::core;
  using namespace cast;
  using namespace cast::cdl;

  void
  BindingWorkingMemoryReader::start() {
    
    m_haveAddr = false;

    // if the set of possible union configurations has been updated,
    // update the monitor accordingly
    
    addChangeFilter(createGlobalTypeFilter<UnionConfiguration>(cdl::ADD),
		    new MemberFunctionChangeReceiver<BindingWorkingMemoryReader>(this,
			&BindingWorkingMemoryReader::fetchThenExtract));

    addChangeFilter(createGlobalTypeFilter<UnionConfiguration>(cdl::OVERWRITE),
		    new MemberFunctionChangeReceiver<BindingWorkingMemoryReader>(this,
			&BindingWorkingMemoryReader::fetchThenExtract));
  }
  
  void 
  BindingWorkingMemoryReader::fetchThenExtract(const cast::cdl::WorkingMemoryChange & _wmc) {
      addrFetchThenExtract(_wmc.address);
      m_currentUnionsAddr = _wmc.address;
      m_haveAddr = true;
  }

  void 
  BindingWorkingMemoryReader::addrFetchThenExtract(const cast::cdl::WorkingMemoryAddress & _addr) {
    try {
      UnionConfigurationPtr config = 
	getMemoryEntry<UnionConfiguration>(_addr);
      extractUnionsFromConfig(config);
    }
    catch (CASTException & e) {
      println("EXCEPTION BindingWorkingMemoryReader::fetchThenExtract: %s", e.message.c_str());
    }
  } 

  void BindingWorkingMemoryReader::extractUnionsFromConfig (UnionConfigurationPtr _config) {

    m_currentUnions.clear();
    for (UnionSequence::iterator i = _config->includedUnions.begin(); 
	 i < _config->includedUnions.end() ; ++i) {
	UnionPtr uni = *i;
	m_currentUnions.insert(pair<string, UnionPtr>(uni->entityID, uni));
	debug("Union ID %s is part of the current configuration", uni->entityID.c_str());
    }

    log("Wow, I currently have %d unions", m_currentUnions.size());
  }

}


